Docker or Podman required for tests since 3.0.0.Alpha6

### Describe the bug

Since 3.0.0.Alpha6 it's not possible to run tests that do not require testcontainers without having docker or podman installed. Instead you get:

```
SomeTest > initializationError FAILED
	    java.lang.NoClassDefFoundError at BasicLoggingEnabler.java:109
	        Caused by: java.lang.ExceptionInInitializerError at ContainerRuntimeUtil.java:86
 ```

### Expected behavior

You should only need podman/docker if testcontainer needs to be startet

### Actual behavior

You need podman/docker even if you only have unit tests without any interaction with services

### How to Reproduce?

1. Have a computer without docker or podman installed (or simply not in Path)
2. Init a simple quarkus app with 3.0.0.Alpha6 without and extensions
3. Try to run the tests

### Output of `uname -a` or `ver`

5.19.0-35-generic

### Output of `java -version`

openjdk version "17.0.6" 2023-01-17 LTS
OpenJDK Runtime Environment Corretto-17.0.6.10.1 (build 17.0.6+10-LTS)
OpenJDK 64-Bit Server VM Corretto-17.0.6.10.1 (build 17.0.6+10-LTS, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.0.0.Alpha6

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

------------------------------------------------------------
Gradle 8.0.2
------------------------------------------------------------

Build time:   2023-03-03 16:41:37 UTC
Revision:     7d6581558e226a580d91d399f7dfb9e3095c2b1d

Kotlin:       1.8.10
Groovy:       3.0.13
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          11.0.18 (Ubuntu 11.0.18+10-post-Ubuntu-0ubuntu122.04)
OS:           Linux 5.19.0-35-generic amd64


### Additional information

Noticed it on the build server first, then tried uninstalling docker + podman locally and that let me reproduce it.

```
build	16-Mar-2023 09:25:58	> Task :project:test
build	16-Mar-2023 09:25:58	
build	16-Mar-2023 09:25:58	Test2 > initializationError FAILED
build	16-Mar-2023 09:25:58	    java.lang.NoClassDefFoundError at BasicLoggingEnabler.java:109
build	16-Mar-2023 09:25:58	        Caused by: java.lang.ExceptionInInitializerError at ContainerRuntimeUtil.java:86
build	16-Mar-2023 09:25:58	
build	16-Mar-2023 09:25:58	Test1 > initializationError FAILED
build	16-Mar-2023 09:25:58	    java.lang.NoClassDefFoundError at BasicLoggingEnabler.java:109
build	16-Mar-2023 09:25:58	        Caused by: java.lang.ExceptionInInitializerError at ContainerRuntimeUtil.java:86
error	16-Mar-2023 09:25:58	
error	16-Mar-2023 09:25:58	31 tests completed, 31 failed
build	16-Mar-2023 09:25:59	
build	16-Mar-2023 09:25:59	> Task :project:test FAILED
```
Was all I got on the ci, with 29 more tests, that all had the exact same error message @gsmet 

On my computer I got the stacktrace

```
Could not initialize class io.quarkus.test.junit.IntegrationTestUtil
java.lang.NoClassDefFoundError: Could not initialize class io.quarkus.test.junit.IntegrationTestUtil
	at io.quarkus.test.junit.BasicLoggingEnabler.beforeAll(BasicLoggingEnabler.java:109)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$12(ClassBasedTestDescriptor.java:395)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:395)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:211)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:84)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:110)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:90)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:85)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy2/jdk.proxy2.$Proxy5.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java
@@ -29,7 +29,7 @@ import io.quarkus.deployment.pkg.builditem.JarBuildItem;
 import io.quarkus.deployment.pkg.builditem.OutputTargetBuildItem;
 import io.quarkus.deployment.steps.MainClassBuildStep;
 import io.quarkus.runtime.LaunchMode;
-import io.quarkus.runtime.util.ContainerRuntimeUtil;
+import io.quarkus.runtime.util.ContainerRuntimeUtil.ContainerRuntime;
 import io.quarkus.utilities.JavaBinFinder;
 
 public class AppCDSBuildStep {
@@ -39,7 +39,6 @@ public class AppCDSBuildStep {
     public static final String CLASSES_LIST_FILE_NAME = "classes.lst";
     private static final String CONTAINER_IMAGE_BASE_BUILD_DIR = "/tmp/quarkus";
     private static final String CONTAINER_IMAGE_APPCDS_DIR = CONTAINER_IMAGE_BASE_BUILD_DIR + "/appcds";
-    public static final ContainerRuntimeUtil.ContainerRuntime CONTAINER_RUNTIME = detectContainerRuntime(false);
 
     @BuildStep(onlyIf = AppCDSRequired.class)
     public void requested(OutputTargetBuildItem outputTarget, BuildProducer<AppCDSRequestedBuildItem> producer)
@@ -204,12 +203,10 @@ public class AppCDSBuildStep {
     // generate the classes file on the host
     private List<String> dockerRunCommands(OutputTargetBuildItem outputTarget, String containerImage,
             String containerWorkingDir) {
-        if (CONTAINER_RUNTIME == ContainerRuntimeUtil.ContainerRuntime.UNAVAILABLE) {
-            throw new IllegalStateException("No container runtime was found. "
-                    + "Make sure you have either Docker or Podman installed in your environment.");
-        }
+        ContainerRuntime containerRuntime = detectContainerRuntime(true);
+
         List<String> command = new ArrayList<>(10);
-        command.add(CONTAINER_RUNTIME.getExecutableName());
+        command.add(containerRuntime.getExecutableName());
         command.add("run");
         command.add("-v");
         command.add(outputTarget.getOutputDirectory().toAbsolutePath().toString() + ":" + CONTAINER_IMAGE_BASE_BUILD_DIR

--- a/core/deployment/src/test/java/io/quarkus/deployment/pkg/steps/NativeImageBuildContainerRunnerTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/pkg/steps/NativeImageBuildContainerRunnerTest.java
@@ -1,6 +1,5 @@
 package io.quarkus.deployment.pkg.steps;
 
-import static io.quarkus.deployment.pkg.steps.AppCDSBuildStep.CONTAINER_RUNTIME;
 import static org.assertj.core.api.Assertions.assertThat;
 
 import java.nio.file.Path;
@@ -19,10 +18,8 @@ class NativeImageBuildContainerRunnerTest {
     @DisabledIfSystemProperty(named = "avoid-containers", matches = "true")
     @Test
     void testBuilderImageBeingPickedUp() {
-        if (CONTAINER_RUNTIME == ContainerRuntimeUtil.ContainerRuntime.UNAVAILABLE) {
-            throw new IllegalStateException("No container runtime was found. "
-                    + "Make sure you have either Docker or Podman installed in your environment.");
-        }
+        ContainerRuntimeUtil.ContainerRuntime containerRuntime = ContainerRuntimeUtil.detectContainerRuntime(true);
+
         NativeConfig nativeConfig = new NativeConfig();
         nativeConfig.containerRuntime = Optional.empty();
         boolean found;
@@ -31,7 +28,7 @@ class NativeImageBuildContainerRunnerTest {
 
         nativeConfig.builderImage = "graalvm";
         localRunner = new NativeImageBuildLocalContainerRunner(nativeConfig, Path.of("/tmp"));
-        command = localRunner.buildCommand(CONTAINER_RUNTIME.getExecutableName(), Collections.emptyList(),
+        command = localRunner.buildCommand(containerRuntime.getExecutableName(), Collections.emptyList(),
                 Collections.emptyList());
         found = false;
         for (String part : command) {
@@ -44,7 +41,7 @@ class NativeImageBuildContainerRunnerTest {
 
         nativeConfig.builderImage = "mandrel";
         localRunner = new NativeImageBuildLocalContainerRunner(nativeConfig, Path.of("/tmp"));
-        command = localRunner.buildCommand(CONTAINER_RUNTIME.getExecutableName(), Collections.emptyList(),
+        command = localRunner.buildCommand(containerRuntime.getExecutableName(), Collections.emptyList(),
                 Collections.emptyList());
         found = false;
         for (String part : command) {
@@ -57,7 +54,7 @@ class NativeImageBuildContainerRunnerTest {
 
         nativeConfig.builderImage = "RandomString";
         localRunner = new NativeImageBuildLocalContainerRunner(nativeConfig, Path.of("/tmp"));
-        command = localRunner.buildCommand(CONTAINER_RUNTIME.getExecutableName(), Collections.emptyList(),
+        command = localRunner.buildCommand(containerRuntime.getExecutableName(), Collections.emptyList(),
                 Collections.emptyList());
         found = false;
         for (String part : command) {

--- a/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
@@ -5,6 +5,7 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
+import java.util.concurrent.TimeUnit;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
@@ -16,7 +17,7 @@ import io.smallrye.config.SmallRyeConfig;
 public final class ContainerRuntimeUtil {
 
     private static final Logger log = Logger.getLogger(ContainerRuntimeUtil.class);
-    private static final String DOCKER_EXECUTABLE = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class)
+    private static final String CONTAINER_EXECUTABLE = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class)
             .getOptionalValue("quarkus.native.container-runtime", String.class).orElse(null);
 
     /**
@@ -51,8 +52,8 @@ public final class ContainerRuntimeUtil {
             // podman version 2.1.1
             String podmanVersionOutput;
             boolean podmanAvailable;
-            if (DOCKER_EXECUTABLE != null) {
-                if (DOCKER_EXECUTABLE.trim().equalsIgnoreCase("docker")) {
+            if (CONTAINER_EXECUTABLE != null) {
+                if (CONTAINER_EXECUTABLE.trim().equalsIgnoreCase("docker")) {
                     dockerVersionOutput = getVersionOutputFor(ContainerRuntime.DOCKER);
                     dockerAvailable = dockerVersionOutput.contains("Docker version");
                     if (dockerAvailable) {
@@ -60,7 +61,7 @@ public final class ContainerRuntimeUtil {
                         return ContainerRuntime.DOCKER;
                     }
                 }
-                if (DOCKER_EXECUTABLE.trim().equalsIgnoreCase("podman")) {
+                if (CONTAINER_EXECUTABLE.trim().equalsIgnoreCase("podman")) {
                     podmanVersionOutput = getVersionOutputFor(ContainerRuntime.PODMAN);
                     podmanAvailable = podmanVersionOutput.startsWith("podman version");
                     if (podmanAvailable) {
@@ -125,8 +126,14 @@ public final class ContainerRuntimeUtil {
             final ProcessBuilder pb = new ProcessBuilder(containerRuntime.getExecutableName(), "--version")
                     .redirectErrorStream(true);
             versionProcess = pb.start();
-            versionProcess.waitFor();
-            return new String(versionProcess.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
+            final int timeoutS = 10;
+            if (versionProcess.waitFor(timeoutS, TimeUnit.SECONDS)) {
+                return new String(versionProcess.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
+            } else {
+                log.debugf("Failure. It took command %s more than %d seconds to execute.", containerRuntime.getExecutableName(),
+                        timeoutS);
+                return "";
+            }
         } catch (IOException | InterruptedException e) {
             // If an exception is thrown in the process, just return an empty String
             log.debugf(e, "Failure to read version output from %s", containerRuntime.getExecutableName());

--- a/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
@@ -5,9 +5,6 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.StandardOpenOption;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
@@ -21,16 +18,12 @@ public final class ContainerRuntimeUtil {
     private static final Logger log = Logger.getLogger(ContainerRuntimeUtil.class);
     private static final String DOCKER_EXECUTABLE = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class)
             .getOptionalValue("quarkus.native.container-runtime", String.class).orElse(null);
-    /*
-     * Caching the value in a file helps us only as much as one JVM execution is concerned.
-     * Test suite's pom.xml sets things like <argLine>-Djava.io.tmpdir="${project.build.directory}"</argLine>,
-     * so the file could appear in /tmp/ or C:\Users\karm\AppData\Local\Temp\ or in fact in
-     * quarkus/integration-tests/something/target.
-     * There is no point in reaching it in `Path.of(Paths.get("").toAbsolutePath().toString(), "target",
-     * "quarkus_container_runtime.txt")`
-     * as the file is deleted between JVM executions anyway.
+
+    /**
+     * Static variable is not used because the class gets loaded by different classloaders at
+     * runtime and the container runtime would be detected again and again unnecessarily.
      */
-    static final Path CONTAINER_RUNTIME = Path.of(System.getProperty("java.io.tmpdir"), "quarkus_container_runtime.txt");
+    private static final String CONTAINER_RUNTIME_SYS_PROP = "quarkus-local-container-runtime";
 
     private ContainerRuntimeUtil() {
     }
@@ -94,39 +87,24 @@ public final class ContainerRuntimeUtil {
     }
 
     private static ContainerRuntime loadConfig() {
-        try {
-            if (Files.isReadable(CONTAINER_RUNTIME)) {
-                final String runtime = Files.readString(CONTAINER_RUNTIME, StandardCharsets.UTF_8);
-                if (ContainerRuntime.DOCKER.name().equalsIgnoreCase(runtime)) {
-                    return ContainerRuntime.DOCKER;
-                } else if (ContainerRuntime.PODMAN.name().equalsIgnoreCase(runtime)) {
-                    return ContainerRuntime.PODMAN;
-                } else if (ContainerRuntime.UNAVAILABLE.name().equalsIgnoreCase(runtime)) {
-                    return ContainerRuntime.UNAVAILABLE;
-                } else {
-                    log.warnf("The file %s contains an unknown value %s. Ignoring it.",
-                            CONTAINER_RUNTIME.toAbsolutePath(), runtime);
-                    return null;
-                }
-            } else {
-                return null;
-            }
-        } catch (IOException e) {
-            log.warnf("Error reading file %s. Ignoring it. See: %s",
-                    CONTAINER_RUNTIME.toAbsolutePath(), e);
+        final String runtime = System.getProperty(CONTAINER_RUNTIME_SYS_PROP);
+        if (runtime == null) {
+            return null;
+        } else if (ContainerRuntime.DOCKER.name().equalsIgnoreCase(runtime)) {
+            return ContainerRuntime.DOCKER;
+        } else if (ContainerRuntime.PODMAN.name().equalsIgnoreCase(runtime)) {
+            return ContainerRuntime.PODMAN;
+        } else if (ContainerRuntime.UNAVAILABLE.name().equalsIgnoreCase(runtime)) {
+            return ContainerRuntime.UNAVAILABLE;
+        } else {
+            log.warnf("System property %s contains an unknown value %s. Ignoring it.",
+                    CONTAINER_RUNTIME_SYS_PROP, runtime);
             return null;
         }
     }
 
     private static void storeConfig(ContainerRuntime containerRuntime) {
-        try {
-            Files.writeString(CONTAINER_RUNTIME, containerRuntime.name(), StandardCharsets.UTF_8,
-                    StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
-            CONTAINER_RUNTIME.toFile().deleteOnExit();
-        } catch (IOException e) {
-            log.warnf("Error writing to file %s. Ignoring it. See: %s",
-                    CONTAINER_RUNTIME.toAbsolutePath(), e);
-        }
+        System.setProperty(CONTAINER_RUNTIME_SYS_PROP, containerRuntime.name());
     }
 
     private static String getVersionOutputFor(ContainerRuntime containerRuntime) {

--- a/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
@@ -45,43 +45,55 @@ public final class ContainerRuntimeUtil {
             return containerRuntime;
         } else {
             // Docker version 19.03.14, build 5eb3275d40
-            final String dockerVersionOutput = getVersionOutputFor(ContainerRuntime.DOCKER);
-            boolean dockerAvailable = dockerVersionOutput.contains("Docker version");
+            String dockerVersionOutput;
+            boolean dockerAvailable;
             // Check if Podman is installed
             // podman version 2.1.1
-            final String podmanVersionOutput = getVersionOutputFor(ContainerRuntime.PODMAN);
-            boolean podmanAvailable = podmanVersionOutput.startsWith("podman version");
+            String podmanVersionOutput;
+            boolean podmanAvailable;
             if (DOCKER_EXECUTABLE != null) {
-                if (DOCKER_EXECUTABLE.trim().equalsIgnoreCase("docker") && dockerAvailable) {
-                    storeConfig(ContainerRuntime.DOCKER);
-                    return ContainerRuntime.DOCKER;
-                } else if (DOCKER_EXECUTABLE.trim().equalsIgnoreCase("podman") && podmanAvailable) {
-                    storeConfig(ContainerRuntime.PODMAN);
-                    return ContainerRuntime.PODMAN;
-                } else {
-                    log.warn("quarkus.native.container-runtime config property must be set to either podman or docker " +
-                            "and the executable must be available. Ignoring it.");
+                if (DOCKER_EXECUTABLE.trim().equalsIgnoreCase("docker")) {
+                    dockerVersionOutput = getVersionOutputFor(ContainerRuntime.DOCKER);
+                    dockerAvailable = dockerVersionOutput.contains("Docker version");
+                    if (dockerAvailable) {
+                        storeConfig(ContainerRuntime.DOCKER);
+                        return ContainerRuntime.DOCKER;
+                    }
                 }
+                if (DOCKER_EXECUTABLE.trim().equalsIgnoreCase("podman")) {
+                    podmanVersionOutput = getVersionOutputFor(ContainerRuntime.PODMAN);
+                    podmanAvailable = podmanVersionOutput.startsWith("podman version");
+                    if (podmanAvailable) {
+                        storeConfig(ContainerRuntime.PODMAN);
+                        return ContainerRuntime.PODMAN;
+                    }
+                }
+                log.warn("quarkus.native.container-runtime config property must be set to either podman or docker " +
+                        "and the executable must be available. Ignoring it.");
             }
+            dockerVersionOutput = getVersionOutputFor(ContainerRuntime.DOCKER);
+            dockerAvailable = dockerVersionOutput.contains("Docker version");
             if (dockerAvailable) {
                 // Check if "docker" is an alias to "podman"
-                if (dockerVersionOutput.equals(podmanVersionOutput)) {
+                if (dockerVersionOutput.startsWith("podman version")) {
                     storeConfig(ContainerRuntime.PODMAN);
                     return ContainerRuntime.PODMAN;
                 }
                 storeConfig(ContainerRuntime.DOCKER);
                 return ContainerRuntime.DOCKER;
-            } else if (podmanAvailable) {
+            }
+            podmanVersionOutput = getVersionOutputFor(ContainerRuntime.PODMAN);
+            podmanAvailable = podmanVersionOutput.startsWith("podman version");
+            if (podmanAvailable) {
                 storeConfig(ContainerRuntime.PODMAN);
                 return ContainerRuntime.PODMAN;
+            }
+            if (required) {
+                throw new IllegalStateException("No container runtime was found. "
+                        + "Make sure you have either Docker or Podman installed in your environment.");
             } else {
-                if (required) {
-                    throw new IllegalStateException("No container runtime was found. "
-                            + "Make sure you have either Docker or Podman installed in your environment.");
-                } else {
-                    storeConfig(ContainerRuntime.UNAVAILABLE);
-                    return ContainerRuntime.UNAVAILABLE;
-                }
+                storeConfig(ContainerRuntime.UNAVAILABLE);
+                return ContainerRuntime.UNAVAILABLE;
             }
         }
     }

--- a/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
@@ -41,7 +41,7 @@ public final class ContainerRuntimeUtil {
     }
 
     public static ContainerRuntime detectContainerRuntime(boolean required) {
-        final ContainerRuntime containerRuntime = loadConfig();
+        final ContainerRuntime containerRuntime = loadContainerRuntimeFromSystemProperty();
         if (containerRuntime != null) {
             return containerRuntime;
         } else {
@@ -57,7 +57,7 @@ public final class ContainerRuntimeUtil {
                     dockerVersionOutput = getVersionOutputFor(ContainerRuntime.DOCKER);
                     dockerAvailable = dockerVersionOutput.contains("Docker version");
                     if (dockerAvailable) {
-                        storeConfig(ContainerRuntime.DOCKER);
+                        storeContainerRuntimeInSystemProperty(ContainerRuntime.DOCKER);
                         return ContainerRuntime.DOCKER;
                     }
                 }
@@ -65,7 +65,7 @@ public final class ContainerRuntimeUtil {
                     podmanVersionOutput = getVersionOutputFor(ContainerRuntime.PODMAN);
                     podmanAvailable = podmanVersionOutput.startsWith("podman version");
                     if (podmanAvailable) {
-                        storeConfig(ContainerRuntime.PODMAN);
+                        storeContainerRuntimeInSystemProperty(ContainerRuntime.PODMAN);
                         return ContainerRuntime.PODMAN;
                     }
                 }
@@ -77,46 +77,48 @@ public final class ContainerRuntimeUtil {
             if (dockerAvailable) {
                 // Check if "docker" is an alias to "podman"
                 if (dockerVersionOutput.startsWith("podman version")) {
-                    storeConfig(ContainerRuntime.PODMAN);
+                    storeContainerRuntimeInSystemProperty(ContainerRuntime.PODMAN);
                     return ContainerRuntime.PODMAN;
                 }
-                storeConfig(ContainerRuntime.DOCKER);
+                storeContainerRuntimeInSystemProperty(ContainerRuntime.DOCKER);
                 return ContainerRuntime.DOCKER;
             }
             podmanVersionOutput = getVersionOutputFor(ContainerRuntime.PODMAN);
             podmanAvailable = podmanVersionOutput.startsWith("podman version");
             if (podmanAvailable) {
-                storeConfig(ContainerRuntime.PODMAN);
+                storeContainerRuntimeInSystemProperty(ContainerRuntime.PODMAN);
                 return ContainerRuntime.PODMAN;
             }
+
+            storeContainerRuntimeInSystemProperty(ContainerRuntime.UNAVAILABLE);
+
             if (required) {
                 throw new IllegalStateException("No container runtime was found. "
                         + "Make sure you have either Docker or Podman installed in your environment.");
-            } else {
-                storeConfig(ContainerRuntime.UNAVAILABLE);
-                return ContainerRuntime.UNAVAILABLE;
             }
+
+            return ContainerRuntime.UNAVAILABLE;
         }
     }
 
-    private static ContainerRuntime loadConfig() {
+    private static ContainerRuntime loadContainerRuntimeFromSystemProperty() {
         final String runtime = System.getProperty(CONTAINER_RUNTIME_SYS_PROP);
+
         if (runtime == null) {
             return null;
-        } else if (ContainerRuntime.DOCKER.name().equalsIgnoreCase(runtime)) {
-            return ContainerRuntime.DOCKER;
-        } else if (ContainerRuntime.PODMAN.name().equalsIgnoreCase(runtime)) {
-            return ContainerRuntime.PODMAN;
-        } else if (ContainerRuntime.UNAVAILABLE.name().equalsIgnoreCase(runtime)) {
-            return ContainerRuntime.UNAVAILABLE;
-        } else {
+        }
+
+        ContainerRuntime containerRuntime = ContainerRuntime.valueOf(runtime);
+
+        if (containerRuntime == null) {
             log.warnf("System property %s contains an unknown value %s. Ignoring it.",
                     CONTAINER_RUNTIME_SYS_PROP, runtime);
-            return null;
         }
+
+        return containerRuntime;
     }
 
-    private static void storeConfig(ContainerRuntime containerRuntime) {
+    private static void storeContainerRuntimeInSystemProperty(ContainerRuntime containerRuntime) {
         System.setProperty(CONTAINER_RUNTIME_SYS_PROP, containerRuntime.name());
     }
 
@@ -197,6 +199,10 @@ public final class ContainerRuntimeUtil {
         private Boolean rootless;
 
         public String getExecutableName() {
+            if (this == UNAVAILABLE) {
+                throw new IllegalStateException("Cannot get an executable name when no container runtime is available");
+            }
+
             return this.name().toLowerCase();
         }
 
@@ -212,6 +218,16 @@ public final class ContainerRuntimeUtil {
                 }
             }
             return rootless;
+        }
+
+        public static ContainerRuntime of(String value) {
+            for (ContainerRuntime containerRuntime : values()) {
+                if (containerRuntime.name().equalsIgnoreCase(value)) {
+                    return containerRuntime;
+                }
+            }
+
+            return null;
         }
     }
 }

--- a/extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerConfig.java
+++ b/extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerConfig.java
@@ -50,9 +50,10 @@ public class DockerConfig {
 
     /**
      * Name of binary used to execute the docker commands.
+     * This setting can override the global container runtime detection.
      */
-    @ConfigItem(defaultValue = "docker")
-    public String executableName;
+    @ConfigItem
+    public Optional<String> executableName;
 
     /**
      * Configuration for Docker Buildx options

--- a/extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerProcessor.java
+++ b/extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerProcessor.java
@@ -204,7 +204,7 @@ public class DockerProcessor {
             boolean buildSuccessful = ExecUtil.exec(out.getOutputDirectory().toFile(), reader, executableName,
                     dockerArgs);
             if (!buildSuccessful) {
-                throw dockerException(dockerArgs);
+                throw dockerException(executableName, dockerArgs);
             }
 
             dockerConfig.buildx.platform
@@ -250,7 +250,8 @@ public class DockerProcessor {
                     containerImageConfig.username.get(),
                     "-p" + containerImageConfig.password.get());
             if (!loginSuccessful) {
-                throw dockerException(new String[] { "-u", containerImageConfig.username.get(), "-p", "********" });
+                throw dockerException(executableName,
+                        new String[] { "-u", containerImageConfig.username.get(), "-p", "********" });
             }
         }
     }
@@ -326,7 +327,7 @@ public class DockerProcessor {
             String[] tagArgs = { "tag", image, additionalTag };
             boolean tagSuccessful = ExecUtil.exec(executableName, tagArgs);
             if (!tagSuccessful) {
-                throw dockerException(tagArgs);
+                throw dockerException(executableName, tagArgs);
             }
         }
     }
@@ -336,14 +337,15 @@ public class DockerProcessor {
         String[] pushArgs = { "push", image };
         boolean pushSuccessful = ExecUtil.exec(executableName, pushArgs);
         if (!pushSuccessful) {
-            throw dockerException(pushArgs);
+            throw dockerException(executableName, pushArgs);
         }
         log.info("Successfully pushed docker image " + image);
     }
 
-    private RuntimeException dockerException(String[] dockerArgs) {
+    private RuntimeException dockerException(String executableName, String[] dockerArgs) {
         return new RuntimeException(
-                "Execution of 'docker " + String.join(" ", dockerArgs) + "' failed. See docker output for more details");
+                "Execution of '" + executableName + " " + String.join(" ", dockerArgs)
+                        + "' failed. See docker output for more details");
     }
 
     private DockerfilePaths getDockerfilePaths(DockerConfig dockerConfig, boolean forNative,

--- a/extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerProcessor.java
+++ b/extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerProcessor.java
@@ -4,6 +4,7 @@ package io.quarkus.container.image.docker.deployment;
 import static io.quarkus.container.image.deployment.util.EnablementUtil.buildContainerImageNeeded;
 import static io.quarkus.container.image.deployment.util.EnablementUtil.pushContainerImageNeeded;
 import static io.quarkus.container.util.PathsUtil.findMainSourcesRoot;
+import static io.quarkus.runtime.util.ContainerRuntimeUtil.detectContainerRuntime;
 
 import java.io.BufferedReader;
 import java.io.IOException;
@@ -87,9 +88,10 @@ public class DockerProcessor {
             throw new RuntimeException("Unable to build docker image. Please check your docker installation");
         }
 
-        var dockerfilePaths = getDockerfilePaths(dockerConfig, false, packageConfig, out);
-        var dockerFileBaseInformationProvider = DockerFileBaseInformationProvider.impl();
-        var dockerFileBaseInformation = dockerFileBaseInformationProvider.determine(dockerfilePaths.getDockerfilePath());
+        DockerfilePaths dockerfilePaths = getDockerfilePaths(dockerConfig, false, packageConfig, out);
+        DockerFileBaseInformationProvider dockerFileBaseInformationProvider = DockerFileBaseInformationProvider.impl();
+        Optional<DockerFileBaseInformationProvider.DockerFileBaseInformation> dockerFileBaseInformation = dockerFileBaseInformationProvider
+                .determine(dockerfilePaths.getDockerfilePath());
 
         if ((compiledJavaVersion.getJavaVersion().isJava17OrHigher() == CompiledJavaVersionBuildItem.JavaVersion.Status.TRUE)
                 && dockerFileBaseInformation.isPresent() && (dockerFileBaseInformation.get().getJavaVersion() < 17)) {
@@ -168,11 +170,11 @@ public class DockerProcessor {
             boolean pushContainerImage,
             PackageConfig packageConfig) {
 
-        var useBuildx = dockerConfig.buildx.useBuildx();
+        boolean useBuildx = dockerConfig.buildx.useBuildx();
 
-        // useBuildx: Whether or not any of the buildx parameters are set
+        // useBuildx: Whether any of the buildx parameters are set
         //
-        // pushImages: Whether or not the user requested the built images to be pushed to a registry
+        // pushImages: Whether the user requested the built images to be pushed to a registry
         // Pushing images is different based on if you're using buildx or not.
         // If not using any of the buildx params (useBuildx == false), then the flow is as it was before:
         //
@@ -196,9 +198,10 @@ public class DockerProcessor {
         }
 
         if (buildContainerImage) {
-            log.infof("Executing the following command to build docker image: '%s %s'", dockerConfig.executableName,
+            final String executableName = dockerConfig.executableName.orElse(detectContainerRuntime(true).getExecutableName());
+            log.infof("Executing the following command to build docker image: '%s %s'", executableName,
                     String.join(" ", dockerArgs));
-            boolean buildSuccessful = ExecUtil.exec(out.getOutputDirectory().toFile(), reader, dockerConfig.executableName,
+            boolean buildSuccessful = ExecUtil.exec(out.getOutputDirectory().toFile(), reader, executableName,
                     dockerArgs);
             if (!buildSuccessful) {
                 throw dockerException(dockerArgs);
@@ -234,7 +237,7 @@ public class DockerProcessor {
 
     private void loginToRegistryIfNeeded(ContainerImageConfig containerImageConfig,
             ContainerImageInfoBuildItem containerImageInfo, DockerConfig dockerConfig) {
-        var registry = containerImageInfo.getRegistry()
+        String registry = containerImageInfo.getRegistry()
                 .orElseGet(() -> {
                     log.info("No container image registry was set, so 'docker.io' will be used");
                     return "docker.io";
@@ -242,7 +245,8 @@ public class DockerProcessor {
 
         // Check if we need to login first
         if (containerImageConfig.username.isPresent() && containerImageConfig.password.isPresent()) {
-            boolean loginSuccessful = ExecUtil.exec(dockerConfig.executableName, "login", registry, "-u",
+            final String executableName = dockerConfig.executableName.orElse(detectContainerRuntime(true).getExecutableName());
+            boolean loginSuccessful = ExecUtil.exec(executableName, "login", registry, "-u",
                     containerImageConfig.username.get(),
                     "-p" + containerImageConfig.password.get());
             if (!loginSuccessful) {
@@ -254,15 +258,17 @@ public class DockerProcessor {
     private String[] getDockerArgs(String image, DockerfilePaths dockerfilePaths, ContainerImageConfig containerImageConfig,
             DockerConfig dockerConfig, ContainerImageInfoBuildItem containerImageInfo, boolean pushImages) {
         List<String> dockerArgs = new ArrayList<>(6 + dockerConfig.buildArgs.size());
-        var useBuildx = dockerConfig.buildx.useBuildx();
+        boolean useBuildx = dockerConfig.buildx.useBuildx();
 
         if (useBuildx) {
+            final String executableName = dockerConfig.executableName.orElse(detectContainerRuntime(true).getExecutableName());
             // Check the executable. If not 'docker', then fail the build
-            if (!DOCKER.equals(dockerConfig.executableName)) {
+            if (!DOCKER.equals(executableName)) {
                 throw new IllegalArgumentException(
                         String.format(
-                                "The 'buildx' properties are specific to 'executable-name=docker' and can not be used with the '%s' executable name. Either remove the `buildx` properties or the `executable-name` property.",
-                                dockerConfig.executableName));
+                                "The 'buildx' properties are specific to 'executable-name=docker' and can not be used with " +
+                                        "the '%s' executable name. Either remove the `buildx` properties or the `executable-name` property.",
+                                executableName));
             }
 
             dockerArgs.add("buildx");
@@ -315,9 +321,10 @@ public class DockerProcessor {
     }
 
     private void createAdditionalTags(String image, List<String> additionalImageTags, DockerConfig dockerConfig) {
+        final String executableName = dockerConfig.executableName.orElse(detectContainerRuntime(true).getExecutableName());
         for (String additionalTag : additionalImageTags) {
             String[] tagArgs = { "tag", image, additionalTag };
-            boolean tagSuccessful = ExecUtil.exec(dockerConfig.executableName, tagArgs);
+            boolean tagSuccessful = ExecUtil.exec(executableName, tagArgs);
             if (!tagSuccessful) {
                 throw dockerException(tagArgs);
             }
@@ -325,8 +332,9 @@ public class DockerProcessor {
     }
 
     private void pushImage(String image, DockerConfig dockerConfig) {
+        final String executableName = dockerConfig.executableName.orElse(detectContainerRuntime(true).getExecutableName());
         String[] pushArgs = { "push", image };
-        boolean pushSuccessful = ExecUtil.exec(dockerConfig.executableName, pushArgs);
+        boolean pushSuccessful = ExecUtil.exec(executableName, pushArgs);
         if (!pushSuccessful) {
             throw dockerException(pushArgs);
         }
@@ -464,7 +472,7 @@ public class DockerProcessor {
                     : mainSourcesRoot.getValue().resolve(dockerfilePath);
             if (!effectiveDockerfilePath.toFile().exists()) {
                 throw new IllegalArgumentException(
-                        "Specified Dockerfile path " + effectiveDockerfilePath.toAbsolutePath().toString() + " does not exist");
+                        "Specified Dockerfile path " + effectiveDockerfilePath.toAbsolutePath() + " does not exist");
             }
             return new ProvidedDockerfile(
                     effectiveDockerfilePath,

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/IntegrationTestUtil.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/IntegrationTestUtil.java
@@ -52,6 +52,7 @@ import io.quarkus.deployment.builditem.DevServicesLauncherConfigResultBuildItem;
 import io.quarkus.paths.PathList;
 import io.quarkus.runtime.configuration.ProfileManager;
 import io.quarkus.runtime.logging.LoggingSetupRecorder;
+import io.quarkus.runtime.util.ContainerRuntimeUtil;
 import io.quarkus.test.common.ArtifactLauncher;
 import io.quarkus.test.common.LauncherUtil;
 import io.quarkus.test.common.PathTestHelper;
@@ -64,7 +65,7 @@ public final class IntegrationTestUtil {
     public static final int DEFAULT_PORT = 8081;
     public static final int DEFAULT_HTTPS_PORT = 8444;
     public static final long DEFAULT_WAIT_TIME_SECONDS = 60;
-    private static final String DOCKER_BINARY = detectContainerRuntime().getExecutableName();
+    public static final ContainerRuntimeUtil.ContainerRuntime CONTAINER_RUNTIME = detectContainerRuntime(false);
 
     private IntegrationTestUtil() {
     }
@@ -336,9 +337,15 @@ public final class IntegrationTestUtil {
     private static void createNetworkIfNecessary(
             final ArtifactLauncher.InitContext.DevServicesLaunchResult devServicesLaunchResult) {
         if (devServicesLaunchResult.manageNetwork() && (devServicesLaunchResult.networkId() != null)) {
+            if (CONTAINER_RUNTIME == ContainerRuntimeUtil.ContainerRuntime.UNAVAILABLE) {
+                throw new IllegalStateException("No container runtime was found. "
+                        + "Make sure you have either Docker or Podman installed in your environment.");
+            }
             try {
                 int networkCreateResult = new ProcessBuilder().redirectError(DISCARD).redirectOutput(DISCARD)
-                        .command(DOCKER_BINARY, "network", "create", devServicesLaunchResult.networkId()).start().waitFor();
+                        .command(CONTAINER_RUNTIME.getExecutableName(), "network", "create",
+                                devServicesLaunchResult.networkId())
+                        .start().waitFor();
                 if (networkCreateResult > 0) {
                     throw new RuntimeException("Creating container network '" + devServicesLaunchResult.networkId()
                             + "' completed unsuccessfully");
@@ -349,7 +356,9 @@ public final class IntegrationTestUtil {
                     public void run() {
                         try {
                             new ProcessBuilder().redirectError(DISCARD).redirectOutput(DISCARD)
-                                    .command(DOCKER_BINARY, "network", "rm", devServicesLaunchResult.networkId()).start()
+                                    .command(CONTAINER_RUNTIME.getExecutableName(), "network", "rm",
+                                            devServicesLaunchResult.networkId())
+                                    .start()
                                     .waitFor();
                         } catch (InterruptedException | IOException ignored) {
                             System.out.println(

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/IntegrationTestUtil.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/IntegrationTestUtil.java
@@ -65,7 +65,6 @@ public final class IntegrationTestUtil {
     public static final int DEFAULT_PORT = 8081;
     public static final int DEFAULT_HTTPS_PORT = 8444;
     public static final long DEFAULT_WAIT_TIME_SECONDS = 60;
-    public static final ContainerRuntimeUtil.ContainerRuntime CONTAINER_RUNTIME = detectContainerRuntime(false);
 
     private IntegrationTestUtil() {
     }
@@ -337,13 +336,11 @@ public final class IntegrationTestUtil {
     private static void createNetworkIfNecessary(
             final ArtifactLauncher.InitContext.DevServicesLaunchResult devServicesLaunchResult) {
         if (devServicesLaunchResult.manageNetwork() && (devServicesLaunchResult.networkId() != null)) {
-            if (CONTAINER_RUNTIME == ContainerRuntimeUtil.ContainerRuntime.UNAVAILABLE) {
-                throw new IllegalStateException("No container runtime was found. "
-                        + "Make sure you have either Docker or Podman installed in your environment.");
-            }
+            ContainerRuntimeUtil.ContainerRuntime containerRuntime = detectContainerRuntime(true);
+
             try {
                 int networkCreateResult = new ProcessBuilder().redirectError(DISCARD).redirectOutput(DISCARD)
-                        .command(CONTAINER_RUNTIME.getExecutableName(), "network", "create",
+                        .command(containerRuntime.getExecutableName(), "network", "create",
                                 devServicesLaunchResult.networkId())
                         .start().waitFor();
                 if (networkCreateResult > 0) {
@@ -356,7 +353,7 @@ public final class IntegrationTestUtil {
                     public void run() {
                         try {
                             new ProcessBuilder().redirectError(DISCARD).redirectOutput(DISCARD)
-                                    .command(CONTAINER_RUNTIME.getExecutableName(), "network", "rm",
+                                    .command(containerRuntime.getExecutableName(), "network", "rm",
                                             devServicesLaunchResult.networkId())
                                     .start()
                                     .waitFor();

